src/client/address_book.cc

```diff
@@ -55,221 +55,292 @@
+ *
+ * NARRATIVE:
+ *
+ * 1. A trusted publisher publishes a subscription
+ * 2. Subscription contains spec-defined host=base64 pairing; one host per line
+ * 3. Kovri checks if we have a list of publishers; if not, uses default
+ * 4. Kovri hooks its subscriber to into an asio timer that regularly
+ *    updates a subscription (only downloads new subscription if Etag is set)
+ * 5. If available, kovri loads default packaged subscription before downloading
+ * 6. Kovri's subscriber checks if downloads subscription/updated subscription
+ * 7. Kovri saves subscription to storage
+ * 8. Kovri repeats download ad infinitum with a timer based on specified constants
+ */
 
-AddressBookFilesystemStorage::AddressBookFilesystemStorage() {
-  auto path = GetPath();
-  if (!boost::filesystem::exists(path)) {
-    // Create directory is necessary
-    if (!boost::filesystem::create_directory(path))
-      LogPrint(eLogError,
-          "AddressBookFilesystemStorage: failed to create addressbook directory");
+void AddressBook::Start(
+    std::shared_ptr<ClientDestination> local_destination) {
+  // We need tunnels so we can download in-net
+  if (!local_destination) {
+    LogPrint(eLogError,
+      "AddressBook: won't start: we need a client destination");
+    return;
+  } else {
+    m_SharedLocalDestination = local_destination;
+    // Load publishers for us to subscribe to
+    LoadPublishers();
+    m_SubscriberUpdateTimer =
+      std::make_unique<boost::asio::deadline_timer>(
+          m_SharedLocalDestination->GetService());
+    m_SubscriberUpdateTimer->expires_from_now(
+        boost::posix_time::minutes(
+            static_cast<std::uint16_t>(SubscriberTimeout::InitialUpdate)));
+    m_SubscriberUpdateTimer->async_wait(
+        std::bind(
+            &AddressBook::SubscriberUpdateTimer,
+            this,
+            std::placeholders::_1));
   }
 }
 
-bool AddressBookFilesystemStorage::GetAddress(
-    const i2p::data::IdentHash& ident,
-    i2p::data::IdentityEx& address) const {
-  auto filename = GetPath() / (ident.ToBase32() + ".b32");
-  std::ifstream f(filename.string(), std::ifstream::binary);
-  if (f.is_open()) {
-    f.seekg(0, std::ios::end);
-    size_t len = f.tellg();
-    if (len < i2p::data::DEFAULT_IDENTITY_SIZE) {
-      LogPrint(eLogError,
-          "AddressBookFilesystemStorage: file ", filename, " is too short. ", len);
-      return false;
+void AddressBook::LoadPublishers() {
+  // TODO(unassigned): this is a one-shot: we won't be able to
+  // edit publisher's file manually with any effect after router start
+  if (m_Subscribers.empty()) {
+    auto publishers = GetDefaultPublishersFilename();
+    LogPrint(eLogInfo, "AddressBook: loading publisher file ", publishers);
+    std::ifstream file(
+        i2p::util::filesystem::GetFullPath(publishers),
+        std::ofstream::in);
+    if (file.is_open()) {
+      std::string publisher;
+      while (!file.eof()) {
```

Comment:
After removing `getline()` from within the block, I think we'd want `while (getline(file, publisher))` or else we'll skip the block entirely.